invenio_github/api.py

```diff
@@ -0,0 +1,337 @@
+from invenio_webhooks.proxies import current_webhooks
+from werkzeug.local import LocalProxy
+from werkzeug.utils import cached_property
+
+from .models import Release, Repository
+from .utils import get_extra_metadata, iso_utcnow
+
+
+class GitHubAPI(object):
+    """Wrapper for GitHub API."""
+
+    def __init__(self, user_id=None):
+        """Create a GitHub API object."""
+        self.user_id = user_id
+
+    @cached_property
+    def api(self):
+        """Return an authenticated GitHub API."""
+        return github3.login(token=self.access_token)
+
+    @cached_property
+    def access_token(self):
+        """Return OAuth access token."""
+        if self.user_id:
+            return RemoteToken.get(
+                self.user_id, self.remote.consumer_key
+            ).access_token
+        return self.remote.get_request_token()[0]
+
+    @property
+    def session_token(self):
+        """Return OAuth session token."""
+        session_token = None
+        if self.user_id is not None:
+            session_token = token_getter(self.remote)
+        if session_token:
+            token = RemoteToken.get(
+                self.user_id, self.remote.consumer_key,
+                access_token=session_token[0]
+            )
+            return token
+        return None
+
+    remote = LocalProxy(
+        lambda: current_oauthclient.oauth.remote_apps[
+            current_app.config['GITHUB_WEBHOOK_RECEIVER_ID']
+        ]
+    )
+    """Return OAuth remote application."""
+
+    @cached_property
+    def account(self):
+        """Return remote account."""
+        return RemoteAccount.get(self.user_id, self.remote.consumer_key)
+
+    def init_account(self):
+        """Setup a new GitHub account."""
+        ghuser = self.api.user()
+        # Setup local access tokens
+        hook_token = ProviderToken.create_personal(
+            'github-webhook',
+            self.user_id,
+            scopes=['webhooks:event'],
+            is_internal=True,
+        )
+        # Initial structure of extra data
+        self.account.extra_data = dict(
+            login=ghuser.login,
+            name=ghuser.name,
+            tokens=dict(
+                webhook=hook_token.id,
+            ),
+            repos=dict(),
+            last_sync=iso_utcnow(),
+        )
+        db.session.add(self.account)
+        # Fetch list of repositories
+        self.sync()
+
+    def sync(self, sync_hooks=True):
+        """Synchronize user repositories."""
+        account = self.account
+        existing_repos = set(account.extra_data.get('repos', {}))
+        new_repos = {}
+
+        repos = self.api.iter_repos(type='all', sort='full_name')
+        repos.headers['Accept'] = 'application/vnd.github.moondragon+json'
+
+        for r in repos:
+            # Next line to be replaced with 'r.permissions['admin']' once
+            # fix has been integrated in github3.py
+            if r.to_json()['permissions']['admin']:
```

Comment:
If we upgrade to the 1.0.0 alpha releases of github3, then this can be changed. See https://github.com/sigmavirus24/github3.py/commit/193fb7613444ad4d25ea31d3e35c5602b5631629